### DISCUSSION TITLE
Imgui on SDL2... possible ?? => Yes, done !

### DISCUSSION THREAD

**Discussion by ryback08** [cavariux/nimgl#63] [Q&A]
Hello Nimgl community & better 2021.

I'm looking to how integrate Imgui on SDL2 (https://github.com/Vladar4/sdl2_nim).

For the moment, I study the nimgl/imgui :
_imgui.nim_ : automatically made by your generator
  _[ imgui ] impl_glfw & impl_opengl_  : totally hand wrapper ? or you have a way to go faster (like c2nim or other way)

So, hopefully if I translate "imgui_impl_sdl.cpp" to impl_sdl.nim in the same way, I can use imgui on sdl2.
Do you think that integration is possible with sdl2_nim ? Do you see another constraint ?

May the Force be with you.

**Comment by lmariscal** [comment]
I believe it is possible, and if you make it I would love for a PR, so it can be integrated into the nimgl/imgui repo.

https://github.com/nimgl/imgui/blob/master/src/imgui/impl_opengl.nim
https://github.com/nimgl/imgui/blob/master/src/imgui/impl_glfw.nim

Those are the implementations for OpenGL and GLFW respectively taken directly from the backends provided by the official ImGui repo (https://github.com/ocornut/imgui/tree/master/backends).
I'm not so knowledgable about SDL, but translating the sdl impl backend from the official repo to the nim bindings should give you the results you are looking for.

**Comment by ryback08** [comment]
I started work here : [https://github.com/ryback08/imgui/blob/master/src/imgui/impl_sdl.nim](https://github.com/ryback08/imgui/blob/master/src/imgui/impl_sdl.nim)
Any help is welcome ;)

**Reply by phargobikcin** [reply]
Hi I came to ask the same question.  I guess my question is there a technical reason not to actually just wrap what is 
in https://github.com/ocornut/imgui/tree/master/backends ?

Thanks.

In the meantime I will look at https://github.com/ryback08/imgui/blob/master/src/imgui/impl_sdl.nim

**Reply by lmariscal** [reply]
@phargobikcin it isn't a wrong approach to just make it work, as long as you make certain the data types being "transmitted" are correct.

The purpose of these implementations made totally in Nim is for them to not only work, but also serve as an example of how to work with the data types provided by NimGL and by the SDL bindings.

**Reply by phargobikcin** [reply]
@lmariscal Thanks.  I am still trying to wrap my head around what is what, I am new to Nim and ImGui.
I will try and flesh out ryback08's code.

**Comment by phargobikcin** [comment]
Hi, I am attempting to get this to work here: 
https://github.com/phargobikcin/nimgl/blob/master/src/nimgl/imgui/impl_sdl.nim

based off of https://github.com/ryback08/imgui/blob/master/src/imgui/impl_sdl.nim.  (Thanks @ryback08!)

**Reply by phargobikcin** [reply]
Have pushed a working version, only tested on linux.

**Reply by ryback08** [reply]
I have a few time. I test your version on windows10 with your NimGL repo & SDL_Nim (from https://github.com/Vladar4/sdl2_nim) in Nim1.6.

I must change all the `import sdl2nim/sdl as sdl` =>` import sdl2_nim/sdl as sdl`

I probably find bug on line 61 in timgui.nim
`igSDL2_ProcessEvent(event)`  =>  `igSDL2_ProcessEvent(e)`  ??

Now, I see a SDL2 windows with opengl and imgui. Great Job 👍 .
I have just a bug if i move the windows :

```
Assertion failed: (g.IO.DeltaTime > 0.0f || g.FrameCount == 0) && "Need a positive DeltaTime!", file imgui/imgui.cpp, line 7262
Traceback (most recent call last)
c:\nim\testgl\exempleNimGl\timgui.nim(110) timgui
c:\nim\testgl\exempleNimGl\timgui.nim(66) main
```

I investigate...

**Reply by ryback08** [reply]
I find the bug. fonction "getTicks" use nim times module : it is not safe for calculate tick !
We can change by std/monotemps nim lib or I prefer use sdl fonction.
Test in win32 ok.

[impl_sdl - test on w10 ok](https://github.com/ryback08/nimgl/blob/master/src/nimgl/imgui/impl_sdl.nim)

**Reply by phargobikcin** [reply]
Great, thanks for testing / fixing.  :)

I guess we can create a pull request?

**Reply by ryback08** [reply]
I guess too. it works for me on W10. for you on Linux.
I am creating a pull request against your version.
 
The next step for me is to add the implementation for opengl2. My target is raspPi.

**Comment by ryback08** [comment]
Hello, 

I begin impl_sdl.nim inspiring of impl_glfw.nim to lmariscal.

For the moment, my implemtation doesn't work.
I have too much misunderstanding to achieve this project (I'm just a hobby programmer).
At the moment, i learn book "Nim in action" to fill my gaps and understand how nim interreact with C.
So, your welcome to this challenge.

**Reply by phargobikcin** [reply]
Thanks for your effort.  I think you did the heavy lifting, I just filled in a few functions to make it work.  :)
